src/core/load.cc

```diff
@@ -178,29 +178,40 @@ CL_DEFUN T_sp core__load_no_package_set(T_sp lsource, T_sp verbose, T_sp print,
       filename = nil<T_O>();
     } else {
       function = nil<T_O>();
-      if (hooks.consp()) {
-        function = oCdr(gc::As<Cons_sp>(hooks)->assoc(pathname->_Type, nil<T_O>(), cl::_sym_equalp, nil<T_O>()));
-        if (function.nilp()) {
-          T_sp stream = cl__open(pathname, kw::_sym_input, ext::_sym_byte8, nil<T_O>(), false, nil<T_O>(), false, external_format,
-                                 nil<T_O>());
-          uint8_t bytes[4];
-          clasp_read_byte8(stream, bytes, 4);
-          cl__close(stream);
-          T_sp magic = clasp_make_fixnum(((uint32_t)bytes[0] << 24) | ((uint32_t)bytes[1] << 16) | ((uint32_t)bytes[2] << 8) |
-                                         ((uint32_t)bytes[3] << 0));
-          function = oCdr(gc::As<Cons_sp>(hooks)->assoc(magic, nil<T_O>(), cl::_sym_equalp, nil<T_O>()));
+      T_sp magic = nil<T_O>();
+      for (T_sp _hooks = hooks; _hooks.notnilp(); _hooks = oCdr(_hooks)) {
+        T_sp key = oCaar(_hooks);
+        if (gc::IsA<String_sp>(key)) {
+          if (cl__equalp(key, pathname->_Type)) {
+            function = oCdar(_hooks);
+            break;
+          }
+        } else {
+          if (magic.nilp()) {
+            T_sp stream = cl__open(pathname, kw::_sym_input, ext::_sym_byte8, nil<T_O>(), false, nil<T_O>(), false, external_format,
+                                   nil<T_O>());
+            uint8_t bytes[4];
+            clasp_read_byte8(stream, bytes, 4);
```

Comment:
Will this signal an error if the file is less than 4 bytes long? I'm not sure it does. Obviously that's a pretty obscure case but I can imagine accidentally trying to load an empty file.

Reply:
No, it will just return what it can read which won't match the 4 byte magic numbers. I tested this with a "touch" to make an empty file.

Reply:
Works